File: MacOS/Type4Me/ASR/Providers/AliyunASRConfig.swift
import Foundation

struct AliyunASRConfig: ASRProviderConfig, Sendable {

    static let provider = ASRProvider.aliyun
    static var displayName: String { L("阿里云", "Alibaba Cloud") }

    static var credentialFields: [CredentialField] {[
        CredentialField(key: "accessKeyId", label: "Access Key ID", placeholder: L("密钥 ID", "Key ID"), isSecure: false, isOptional: false, defaultValue: ""),
        CredentialField(key: "accessKeySecret", label: "Access Key Secret", placeholder: L("密钥", "Secret"), isSecure: true, isOptional: false, defaultValue: ""),
        CredentialField(key: "appKey", label: "App Key", placeholder: L("项目 AppKey", "Project AppKey"), isSecure: false, isOptional: false, defaultValue: ""),
    ]}

    let accessKeyId: String
    let accessKeySecret: String
    let appKey: String

    init?(credentials: [String: String]) {
        guard let kid = credentials["accessKeyId"], !kid.isEmpty,
              let secret = credentials["accessKeySecret"], !secret.isEmpty,
              let appKey = credentials["appKey"], !appKey.isEmpty
        else { return nil }
        self.accessKeyId = kid
        self.accessKeySecret = secret
        self.appKey = appKey
    }

    func toCredentials() -> [String: String] {
        ["accessKeyId": accessKeyId, "accessKeySecret": accessKeySecret, "appKey": appKey]
    }

    var isValid: Bool { !accessKeyId.isEmpty && !accessKeySecret.isEmpty && !appKey.isEmpty }
}

File: Windows/Type4Me/ASR/Providers/AliyunASRConfig.cs
using Type4Me.Localization;

namespace Type4Me.ASR.Providers;

public sealed class AliyunASRConfig
{
    public string AccessKeyId { get; }
    public string AccessKeySecret { get; }
    public string AppKey { get; }

    public static ASRProvider Provider => ASRProvider.Aliyun;
    public static string DisplayName => Loc.L("阿里云", "Alibaba Cloud");

    public static CredentialField[] CredentialFields =>
    [
        new() { Key = "accessKeyId", Label = "Access Key ID", Placeholder = Loc.L("密钥 ID", "Key ID") },
        new() { Key = "accessKeySecret", Label = "Access Key Secret", Placeholder = Loc.L("密钥", "Secret"), IsSecure = true },
        new() { Key = "appKey", Label = "App Key", Placeholder = Loc.L("项目 AppKey", "Project AppKey") },
    ];

    private AliyunASRConfig(string accessKeyId, string accessKeySecret, string appKey)
    {
        AccessKeyId = accessKeyId; AccessKeySecret = accessKeySecret; AppKey = appKey;
    }

    public static AliyunASRConfig? TryCreate(Dictionary<string, string> credentials)
    {
        if (!credentials.TryGetValue("accessKeyId", out var kid) || string.IsNullOrEmpty(kid)) return null;
        if (!credentials.TryGetValue("accessKeySecret", out var secret) || string.IsNullOrEmpty(secret)) return null;
        if (!credentials.TryGetValue("appKey", out var appKey) || string.IsNullOrEmpty(appKey)) return null;
        return new AliyunASRConfig(kid, secret, appKey);
    }

    public Dictionary<string, string> ToCredentials() => new()
    {
        ["accessKeyId"] = AccessKeyId, ["accessKeySecret"] = AccessKeySecret, ["appKey"] = AppKey,
    };

    public bool IsValid => !string.IsNullOrEmpty(AccessKeyId) && !string.IsNullOrEmpty(AccessKeySecret) && !string.IsNullOrEmpty(AppKey);
}
